### PR TITLE
Keep as much parsed state as we can when producing frozen compilations

### PR DESCRIPTION
**Pull request by CyrusNajmabadi**

### CODE REVIEW

**Comment by CyrusNajmabadi** [comment]
@jasonmalinowski @ToddGrun this is ready for review.

File: src/Workspaces/Core/Portable/Workspace/Solution/SolutionCompilationState.CompilationTracker.cs [hunk 1]
```diff
@@ -641,35 +641,41 @@ public ICompilationTracker FreezePartialState(CancellationToken cancellationToke
...
+                return new CompilationTracker(frozenProjectState, frozenState, this.SkeletonReferenceCache.Clone());
 
-                    // If we have an in progress state with no steps, then we're just at the current project state.
-                    InProgressState { PendingTranslationActions: [] } => this.ProjectState,
+                ProjectState GetInitialProjectState(CompilationTrackerState? state)
+                {
+                    return state switch
+                    {
+                        // If we don't have an existing state, then transition to a project state without any data.
+                        null => this.ProjectState,
```
Comment (CyrusNajmabadi) [inline src/Workspaces/Core/Portable/Workspace/Solution/SolutionCompilationState.CompilationTracker.cs:655]: intentionally changes.  teh RemoveAllDocuments is now in GetFrozenCompilationState

File: src/Workspaces/Core/Portable/Workspace/Solution/SolutionCompilationState.CompilationTracker.cs [hunk 2]
```diff
@@ -641,35 +641,41 @@ public ICompilationTracker FreezePartialState(CancellationToken cancellationToke
                 if (state is FinalCompilationTrackerState { IsFrozen: true } finalState)
                     return this;
 
-                var projectState = state switch
```
Comment (CyrusNajmabadi) [inline src/Workspaces/Core/Portable/Workspace/Solution/SolutionCompilationState.CompilationTracker.cs:644]: view with whitespace off.

File: src/Workspaces/Core/Portable/Workspace/Solution/SolutionCompilationState.CompilationTracker.cs [hunk 3]
```diff
@@ -680,14 +686,35 @@ CompilationTrackerState GetFrozenCompilationState()
...
+                        foreach (var (documentId, documentState) in projectState.DocumentStates.States)
+                        {
+                            if (documentState.TryGetSyntaxTree(out var alreadyParsedTree))
+                            {
+                                documentsWithTrees.Add(documentState);
+                                alreadyParsedTrees.Add(alreadyParsedTree);
+                            }
+                        }
+
+                        frozenProjectState = frozenProjectState.AddDocuments(documentsWithTrees.ToImmutableAndClear());
```
[the remark this answers is not in the document]
Comment (CyrusNajmabadi) [reply src/Workspaces/Core/Portable/Workspace/Solution/SolutionCompilationState.CompilationTracker.cs:708]: i'd prefer to drive this based on allocation profiles.  immutable types really help us ensure no mistakes.

**Review by ToddGrun — APPROVED** [review]
:shipit:

File: src/Workspaces/Core/Portable/Workspace/Solution/SolutionCompilationState.CompilationTracker.cs (hunk 1 above)
Comment (jasonmalinowski) [reply src/Workspaces/Core/Portable/Workspace/Solution/SolutionCompilationState.CompilationTracker.cs:655]: If that's the case then update the comment.

Comment (jasonmalinowski) [reply src/Workspaces/Core/Portable/Workspace/Solution/SolutionCompilationState.CompilationTracker.cs:655]: And at this point, I just want to see GetIniitalProjectState combined with the other function. There's now two functions that both have the same switches and you can only understand how this works if I look at both of them at once. It's far more cognitive load to understand the structure than just one bigger method here.

File: src/Workspaces/Core/Portable/Workspace/Solution/SolutionCompilationState.CompilationTracker.cs [hunk 4]
```diff
@@ -680,14 +686,35 @@ CompilationTrackerState GetFrozenCompilationState()
...
+
+                        // However, we may have already parsed some of the documents in this compilation.  For example,
+                        // if we're partway through the logic in BuildInProgressStateFromNoCompilationStateAsync.  If
+                        // so, move those parsed documents over to the new project state so we can preserve as much
+                        // information as possible.
+
+                        using var _1 = ArrayBuilder<DocumentState>.GetInstance(out var documentsWithTrees);
+                        using var _2 = ArrayBuilder<SyntaxTree>.GetInstance(out var alreadyParsedTrees);
+
+                        foreach (var (documentId, documentState) in projectState.DocumentStates.States)
```
Comment (jasonmalinowski) [inline src/Workspaces/Core/Portable/Workspace/Solution/SolutionCompilationState.CompilationTracker.cs:699]:
```suggestion
                        foreach (var (documentId, documentState) in projectState.DocumentStates.GetStatesInCompilationOrder())
```

We don't do a great job ensuring the tree order matters, and I'm not sure I care much, but if we have a convenient helper for it then maybe we should at least use it. :smile:

**Review by jasonmalinowski — APPROVED** [review]

File: src/Workspaces/Core/Portable/Workspace/Solution/SolutionCompilationState.CompilationTracker.cs (hunk 4 above)
Comment (CyrusNajmabadi) [reply src/Workspaces/Core/Portable/Workspace/Solution/SolutionCompilationState.CompilationTracker.cs:699]: >  It's far more cognitive load to understand the structure than just one bigger method here.

will change!

**Pull request merged**